model-engine/model_engine_server/inference/vllm/vllm_batch.py

```diff
@@ -0,0 +1,341 @@
+        logprobs=1 if content.return_token_log_probs else None,
+        presence_penalty=content.presence_penalty or 0.0,
+        frequency_penalty=content.frequency_penalty or 0.0,
+        top_k=content.top_k or -1,
+        top_p=content.top_p or 1.0,
+        skip_special_tokens=(
+            content.skip_special_tokens if content.skip_special_tokens is not None else True
+        ),
+    )
+
+    results_generators: List[AsyncIterator[RequestOutput]] = []
+    for prompt in prompts:
+        request_id = random_uuid()
+        results_generator = engine.generate(
+            prompt,
+            sampling_params=sampling_params,
+            request_id=request_id,
+        )
+        results_generators.append(results_generator)
+
+    return_token_log_probs = True
+
+    generator = merge_async_iterators(*results_generators)
+    outputs: List[Optional[CompletionV1Output]] = [None] * len(prompts)
+    tokens: List[List[TokenOutput]] = [list()] * len(prompts)
+    async for i, res in generator:
+        # There should only be one output
+        output = res.outputs[-1]
+
+        if return_token_log_probs and output.logprobs is not None:
+            # Sometime the logprobs are not present in the output
+            logprobs = output.logprobs[-1]
+            for token_id in logprobs.keys():
+                tokens[i].append(
+                    TokenOutput(
+                        token=logprobs[token_id].decoded_token,
+                        log_prob=logprobs[token_id].logprob,
+                    )
+                )
+
+        if res.finished:
+            outputs[i] = CompletionV1Output(
+                text=output.text,
+                num_prompt_tokens=len(res.prompt_token_ids),
+                num_completion_tokens=len(output.token_ids),
+                tokens=[
+                    token.model_dump() for token in tokens[i]
+                ],  # Not sure why, but pydantic doesn't like when I pass it TokenOutput directly but works when I encode it as a dict...
+            )
+            bar.update(1)
+
+    return outputs
+
+
+async def generate_v2_completions(
+    engine: AsyncEngineClient,
+    requests: Union[List[CompletionRequest], List[ChatCompletionRequest]],
+) -> List[Union[CompletionResponse, ErrorResponse, None]]:
+    bar = tqdm(total=len(requests), desc="Processed requests")
+    results_generators: List[
+        Coroutine[
+            Any,
+            Any,
+            Union[ErrorResponse, AsyncGenerator[str, None], CompletionResponse],
+        ]
+    ] = []
+    for request in requests:
+        if isinstance(request, CompletionRequest):
+            results_generators.append(openai_serving_completion.create_completion(request))
+        elif isinstance(request, ChatCompletionRequest):
+            results_generators.append(openai_serving_chat.create_chat_completion(request))
+        else:
+            assert_never(request)
+
+    results_generator = await_coroutines(*results_generators)
```

Comment:
does this guarantee ordering?

Reply:
the coroutines don't complete in order, but the function itself wraps each generator with the corresponding index in the sequence/list so the output order is fixed